### PR TITLE
Saved recipes persist through rounds

### PR DESCRIPTION
**Pull request by Walarks**
## About The Pull Request

Chemical recipes are saved on the client, meaning they are always available and persist through games. Up to 10 can be saved.

https://user-images.githubusercontent.com/45896482/102711935-fa48d600-42c5-11eb-9ca4-a1742f3251ff.mp4

## Why It's Good For The Game

While making custom mixes is fun, having to input the same combination of chems each round can be tedious. If common usage of chems makes marines too powerful, they should be balanced accordingly.

## Changelog
:cl:Slotbot
tweak: saved chem recipes are persistent
tweak: adds 1u as an option when dispensing chems
/:cl:

### CODE REVIEW

**Comment by tgstation-server** [comment]
You currently have a negative Fix/Feature pull request delta of -4. Maintainers may close this PR at will. Fixing issues or improving the codebase will improve this score.

**Review by Hughgent — APPROVED** [review]
Neat, no TM though as it messes with the save file.

Can you show it working locally please?

**Comment by Walarks** [comment]
> Neat, no TM though as it messes with the save file.
> 
> Can you show it working locally please?

added

**Comment by Hughgent** [comment]
Your video shows 'between machines' I was talking about 'between rounds'

**Comment by Walarks** [comment]
> Your video shows 'between machines' I was talking about 'between rounds'

I posted a new video where the round is restarted.

File: code/modules/reagents/machinery/chem_dispenser.dm [hunk 1]
```diff
@@ -20,6 +20,7 @@
 	var/amount = 30
 	var/recharge_amount = 10
 	var/recharge_counter = 0
+	var/possible_transfer_amounts = list(1,5,10,15,20,30,60)
```
Comment (TiviPlus) [inline code/modules/reagents/machinery/chem_dispenser.dm:23]: Why not static list?

File: code/modules/reagents/machinery/chem_dispenser.dm [hunk 2]
```diff
@@ -133,6 +132,7 @@
 	data["energy"] = cell.charge ? cell.charge * powerefficiency : "0" //To prevent NaN in the UI.
 	data["maxEnergy"] = cell.maxcharge * powerefficiency
 	data["isBeakerLoaded"] = beaker ? 1 : 0
+	data["beakerTransferAmounts"] = possible_transfer_amounts
```
Comment (TiviPlus) [inline code/modules/reagents/machinery/chem_dispenser.dm:135]: Why not static ui data?

**Review by TiviPlus — COMMENTED** [review]
Good otherwise

File: code/modules/reagents/machinery/chem_dispenser.dm (hunk 1 above)
Comment (Walarks) [reply code/modules/reagents/machinery/chem_dispenser.dm]: changed

File: code/modules/reagents/machinery/chem_dispenser.dm (hunk 2 above)
Comment (Walarks) [reply code/modules/reagents/machinery/chem_dispenser.dm]: changed

File: code/modules/reagents/machinery/chem_dispenser.dm [hunk 3]
```diff
@@ -21,6 +21,11 @@
 	var/recharge_amount = 10
 	var/recharge_counter = 0
 
+	///UI data holder
+	var/data = list()
```
Comment (TiviPlus) [inline code/modules/reagents/machinery/chem_dispenser.dm:25]: No.

File: code/modules/reagents/machinery/chem_dispenser.dm [hunk 4]
```diff
@@ -127,8 +130,10 @@
 		ui = new(user, src, ui_key, "ChemDispenser", name, ui_x, ui_y, master_ui, state)
 		ui.open()
 
+/obj/machinery/chem_dispenser/ui_static_data(mob/user)
+	data["beakerTransferAmounts"] = possible_transfer_amounts
```
Comment (TiviPlus) [inline code/modules/reagents/machinery/chem_dispenser.dm:134]: This isnt how you use ui_static_data

**Review by TiviPlus — APPROVED** [review]

**Pull request merged**